[cxx-interop] Flip the switch: only import safe APIs.

zoecarver opened the pull request:
This patch adds is the first of many to flip the switch and stop eagerly importing C++ APIs. We will now only import safe APIs or APIs that are annotated by the user. I will follow this patch up with improved diagnostics, testing, and documentation.

zoecarver committed:
[cxx-interop] Flip the switch: only import safe APIs.

zoecarver commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -4495,6 +4554,18 @@ namespace {
+      if (decl->getReturnType()->isPointerType() ||
+          decl->getReturnType()->isReferenceType()) {
+        CXXMethodBridging bridgingInfo(decl);
+        if (bridgingInfo.getClangName() != "begin" &&
+            bridgingInfo.getClangName() != "end" &&
+            !decl->isOverloadedOperator() && !decl->isStatic() &&
+            !hasUnsafeAPIAttr(decl) && !hasImportAsRefAttr(decl->getParent()))
```
- [ ] add tests for all of these

zoecarver commented on test/Interop/Cxx/class/inheritance/Inputs/functions.h:
```diff
@@ -7,32 +7,48 @@ struct NonTrivial {
-  inline const char *inNonTrivial() const { return "NonTrivial::inNonTrivial"; }
-  inline const char *inNonTrivialWithArgs(int a, int b) const {
+  inline const char *inNonTrivial() const
+      __attribute__((swift_attr("import_unsafe_api"))) {
```
Review note: I was originally update all these tests to be safe, but I decided that should happen in a follow up PR. These are testing very specific things, so we should upgrade them slowly/incrementally to ensure we don't regress.

egorzhdan commented:
This essentially makes the `enable_unsafe_eager_importing` flag viral in a sense that if one of your Swift dependencies uses this flag, you also have to use it even if you don't reference any C++ APIs at all, otherwise you'll get SIL deserialization errors: the decls referenced by that dependency won't be found when compiling the current module.

This makes me think that ClangImporter might not be the right place for this logic. I think a better approach would be to import all APIs in ClangImporter, and then reject unsafe calls during type checking. Let's discuss offline :)

zoecarver commented on test/Interop/Cxx/stdlib/use-std-iterator.swift:
```diff
@@ -1,37 +0,0 @@
-// RUN: %target-run-simple-swift(-I %S/Inputs -Xfrontend -enable-experimental-cxx-interop)
```
This is not something that we want people to be able to do anymore. The correct way to do this is with the stdlib overlay, which we already have tests for.

zoecarver commented on docs/CppInteroperability/UserManual.md:
```diff
@@ -0,0 +1,37 @@
+
+**Unsafe Lifetime Types:** are types that have custom lifetime operations such as custom copy constructors or custom destructors. These custom lifetime operations are assumed to be unsafe in all contexts.
+
+**Un-importable Types:** are types that are not copyable or destructible. These types cannot be represented in Swift, so they cannot be imported (even if they are marked with the `import_unsafe` attribute).
```
Un-importable types are the same as unsafe lifetime types. I should update the definition of unsafe lifetime types and remove this.

zoecarver commented:
@swift-ci please test

egorzhdan commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -5828,3 +5828,295 @@ void ClangImporter::diagnoseMemberValue(const DeclName &name,
+    return CxxRecordSemanticsKind::Owned;
+  }
+
+  if (hasIteratorAPIAttr(decl)) {
```
I don't think we should require the user to annotate all iterators with an attribute. Swift can detect that the type is an iterator statically without an attribute by doing `clangRecordDecl->lookup(getClangASTContext().Idents.get("iterator_category"))` – if there is a subtype called `iterator_category`, then this type is an iterator (see `swift::conformToCxxIteratorIfNeeded` from https://github.com/apple/swift/pull/60001).

If the iterator type doesn't define `iterator_category`, then the user can add an attribute, but I don't think we should require attributes for something we can correctly deduce statically.

zoecarver replied:
As discussed offline: we will use both to determine what's an iterator.

egorzhdan commented on stdlib/public/Cxx/std/String.swift:
```diff
@@ -12,16 +12,19 @@
+    let count = string.count
+    self.init(count, value_type(0))
+    var string = string
+    string.withUTF8 {
```
Thanks for fixing this! 👍

zoecarver replied:
I thought this would be a small perf improvement but actually seems to not work. We can investigate in a separate follow up patch.

egorzhdan commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -5828,3 +5828,295 @@ void ClangImporter::diagnoseMemberValue(const DeclName &name,
+       !copyAssignOperatorIsDefaulted(decl)) ||
+      (decl->hasUserDeclaredDestructor() &&
+       !decl->getDestructor()->isDefaulted()))
+    return false;
```
I'm not sure I understand why a type with a custom copy constructor is unsafe. I think this is probably too restrictive, let's discuss.

zoecarver committed:
[nfc][cxx-interop] Add diagnostics when something cannot be imported.

zoecarver committed:
[nfc] Clang format changes.

zoecarver commented on docs/CppInteroperability/UserManual.md:
```diff
@@ -0,0 +1,37 @@
+The Swift compiler enforces certain API rules, not to ensure a completely safe C++ API interface, but to prevent especially unsafe patterns that will likely lead to bugs. Many of these unsafe patterns stem from the subtly different lifetime semantics in C++ and Swift and aim to prevent memory projections of owned types. The currently enforced rules are as follows:
+
+* Unsafe lifetime types and un-importable types are not allowed to be used in any contexts.
+* Unsafe pointer types are not allowed to represent unsafe memory projections from owned types. Note: unsafe pointer types *are* allowed to represent memory projections from reference types. Note: the Swift compiler assumes that global and static functions do not return projections.
```
We don't have the idea of immortal vs other foreign reference types. Above it clarifies that foreign reference types must be immortal or manually managed, so I think it's fine. This states our current support, not our future plans.

zoecarver commented on test/Interop/Cxx/class/invalid-class-errors.swift:
```diff
@@ -0,0 +1,38 @@
+
+import Test
+
+// CHECK: note: record 'X' is not defined (incomplete)
```
@NuriAmari thanks again for building all the infrastructure for these diagnostics. It made it super easy (one line) to add diagnostics for all these other cases.

NuriAmari replied:
My pleasure, glad it's being put to good use 👍

zoecarver committed:
[cxx-interop] Remove `UnsafeLifetimeOperation` record semantic kind.

This makes "owned" the default kind for records that have custom copy constructors and no pointer-members.

zoecarver commented:
@swift-ci please test.

zoecarver committed:
[cxx-interop] Simplify logic in `IsSafeUseOfCxxDecl`; fix bug for pointer types.

Previously types containing pointers would mark all C++ methods as unsafe.

zoecarver commented:
@swift-ci please test.

zoecarver committed:
Mark a few API uses in Swift Compiler Sources as unsafe.

zoecarver commented:
@swift-ci please test.

zoecarver commented:
@swift-ci please test.

zoecarver committed:
[libswift] Update SWIFT_IMPORT_REFERENCE to use import_reference.

zoecarver commented:
@swift-ci please test.

zoecarver committed:
[cxx-interop] Don't add incomplete records to lookup table.

zoecarver committed:
[nfc][cxx-interop] Fix test failures and diagnostic names.

zoecarver committed:
[cxx-interop][overlay] Revert memcpy string init approach.

For some reason this isn't working, so I reverted to the previous approach. We can try to re-visit this in the future for a potential perf improvement.

zoecarver commented:
@swift-ci please test.

zoecarver commented:
@swift-ci please smoke test linux.

zoecarver commented:
@swift-ci please smoke test linux.

zoecarver commented:
@swift-ci please smoke test linux.

zoecarver commented:
@swift-ci please smoke test linux.

zoecarver committed:
[cxx-interop] Fix predicate to find copy assignment operators.

zoecarver commented:
@swift-ci please smoke test.

zoecarver commented:
@swift-ci please test.
